aws/resource_aws_budgets_budget.go

```diff
@@ -0,0 +1,431 @@
+		accountID = v.(string)
+	} else {
+		accountID = meta.(*AWSClient).accountid
+	}
+
+	budget, err := expandBudgetsBudgetUnmarshal(d)
+	if err != nil {
+		return fmt.Errorf("failed creating budget: %v", err)
+	}
+
+	_, err = client.CreateBudget(&budgets.CreateBudgetInput{
+		AccountId: &accountID,
+		Budget:    budget,
+	})
+	if err != nil {
+		return fmt.Errorf("create budget failed: %v", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", accountID, *budget.BudgetName))
+	return resourceAwsBudgetsBudgetUpdate(d, meta)
+}
+
+func resourceAwsBudgetsBudgetRead(d *schema.ResourceData, meta interface{}) error {
+	accountID, budgetName, err := decodeBudgetsBudgetID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	client := meta.(*AWSClient).budgetconn
+	describeBudgetOutput, err := client.DescribeBudget(&budgets.DescribeBudgetInput{
+		BudgetName: &budgetName,
+		AccountId:  &accountID,
+	})
+	if isAWSErr(err, budgets.ErrCodeNotFoundException, "") {
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("describe budget failed: %v", err)
+	}
+
+	flattenedBudget, err := expandBudgetsBudgetFlatten(describeBudgetOutput.Budget)
+	if err != nil {
+		return fmt.Errorf("failed flattening budget output: %v", err)
+	}
+
+	if _, ok := d.GetOk("name"); ok {
+		d.Set("name", flattenedBudget.name)
+	}
+
+	for k, v := range map[string]interface{}{
+		"account_id":        accountID,
+		"budget_type":       flattenedBudget.budgetType,
+		"time_unit":         flattenedBudget.timeUnit,
+		"cost_filters":      convertCostFiltersToStringMap(flattenedBudget.costFilters),
+		"limit_amount":      flattenedBudget.limitAmount,
+		"limit_unit":        flattenedBudget.limitUnit,
+		"cost_types":        []interface{}{flattenedBudget.costTypes},
+		"time_period_start": flattenedBudget.timePeriodStart.Format("2006-01-02_15:04"),
+		"time_period_end":   flattenedBudget.timePeriodEnd.Format("2006-01-02_15:04"),
+	} {
+		if _, ok := d.GetOk(k); ok {
+			if err := d.Set(k, v); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func resourceAwsBudgetsBudgetUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AWSClient).budgetconn
+	var accountID string
+	if v, ok := d.GetOk("account_id"); ok {
+		accountID = v.(string)
+	} else {
+		accountID = meta.(*AWSClient).accountid
+	}
+
+	budget, err := expandBudgetsBudgetUnmarshal(d)
+	if err != nil {
+		return fmt.Errorf("could not create budget: %v", err)
+	}
+
+	_, err = client.UpdateBudget(&budgets.UpdateBudgetInput{
+		AccountId: &accountID,
+		NewBudget: budget,
+	})
+	if err != nil {
+		return fmt.Errorf("update budget failed: %v", err)
+	}
+
+	return resourceAwsBudgetsBudgetRead(d, meta)
+}
+
+func resourceAwsBudgetsBudgetDelete(d *schema.ResourceData, meta interface{}) error {
+	accountID, budgetName, err := decodeBudgetsBudgetID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	client := meta.(*AWSClient).budgetconn
+	_, err = client.DescribeBudget(&budgets.DescribeBudgetInput{
```

Comment:
We can remove the API call here by moving the `isAWSErr()` check under the `if err != nil` check under the `DeleteBudget` call:

```go
if err != nil {
  if isAWSErr(err, budgets.ErrCodeNotFoundException, "") {
    return nil
  }
  return fmt.Errorf("delete budget failed: %s", err)
}
```